core/src/openid.cpp

```diff
@@ -102,12 +102,100 @@ namespace irods::http::openid
 
 		auto json_res{hit_introspection_endpoint(url_encode_body(args))};
 
-		// Validate access token
+		// Check that the access token is active.
+		// The exact meaninng of an active state may vary between OpenID Providers,
+		// so we should assert the other attributes when available.
 		if (!json_res.at("active").get<bool>()) {
 			logging::warn("{}: Access token is invalid or expired.", __func__);
 			return std::nullopt;
 		}
 
+		// The rest of the members are optional, but validate the ones that do exist.
+		// TODO: Verify that this is actually how this works. I don't think we can check this 'client_id'.
+		if (auto client_id_iter{json_res.find("client_id")}; client_id_iter != std::end(json_res)) {
+			if (client_id_iter->get_ref<const std::string&>() != irods::http::globals::oidc_configuration().at("client_id")) {
+				return std::nullopt;
+			}
+		}
+
+		// While not required, ensure the token type is something we can use.
+		// TODO: Consider removing this part. Knowing the 'token_type' after the server has already
+		// processed the token doesn't do anything for us.
+		if (auto token_type_iter{json_res.find("token_type")}; token_type_iter != std::end(json_res)) {
+
+			// There are only 'bearer' tokens and 'mac' tokens defined
+			if (token_type_iter->get_ref<const std::string&>() != "bearer") {
+				return std::nullopt;
+			}
+		}
```

Comment:
Since we have already 'processed' the token, I don't think knowing whether this a 'bearer' or a 'mac' based token is of interest. We also only process bearer tokens.

Reply:
Removed.